### ISSUE TITLE
Starred articles aren't synchronized from TT-RSS to Newsboat

### ISSUE THREAD

**Issue by asch** [newsboat/newsboat#270]
Hi, it looks like the only one-way synchronization of "starred" articles is implemented in TTRSS support. Is there some fundamental problem in implementing synchronization in TTRSS -> newsboat way or developers just forgot? I can try to patch it, but first I would like to know, if there isn't some fundamental problem developers know about.

**Comment by Minoru** [comment]
Hi! I don't recall any such problems discussed. @noctux, do you know anything about this?

**Comment by asch** [comment]
Also I have found the same issues [1,2] reported at old newsbeuter project. Looks like it wasn't solved.

[1] https://github.com/akrennmair/newsbeuter/issues/466
[2] https://github.com/akrennmair/newsbeuter/issues/321

**Comment by Minoru** [comment]
Oh, akrennmair/newsbeuter#321 is blocked on me for more than two years now. That's a shame. I forgot about it entirely.

Looking at that code now, I can say that:
- [this change](https://github.com/dirb/newsbeuter/commit/723b469412185ae5374109959e802d8adaaa250d#diff-fe4e54b49aa936d63086740264cefc01R651) is okay, but it should be copied to `INSERT` statements too. Every SQL statement in `update_rssitem_unlocked` should set or update `flags` field;
- [this change](https://github.com/dirb/newsbeuter/commit/723b469412185ae5374109959e802d8adaaa250d#diff-9f23acca5027f2b0decf88c9e0fd8aaaR278) is okay too, because it makes `update_flags` work [later on](https://github.com/dirb/newsbeuter/commit/723b469412185ae5374109959e802d8adaaa250d#diff-9f23acca5027f2b0decf88c9e0fd8aaaR322).

I think this should be testeable. `cache.cpp` got a decent test suite already, so if you want to hack on this, @asch, you might want to start with extending that. Then, once you got dirb's changes ported, you can fix the TT-RSS code to set flags (that's be untesteable I guess, at least right now).

**Comment by noctux** [comment]
Hey:

On the technical side, this should be more or less straightforward. The starred status is contained in the restponse to getHeadlines, which we are using anyways (see the "marked" entry below):

```
  {
   "id": 304934,
   "guid": "SHA1:a664aae8c10a044144b387a6c4987a3940fc391a",
   "unread": true,
   "marked": true,
   "published": false,
   "updated": 1536562082,
   "is_updated": false,
   "title": "Felony Disenfranchisement: Last Week Tonight with John Oliver (HBO)",
   "link": "https:\/\/www.youtube.com\/watch?v=NpPyLcQ2vdI",
   "feed_id": "89",
   "tags": [
    ""
   ],
   "content": "",
   "labels": [],
   "feed_title": "LastWeekTonight",
   "comments_count": 0,
   "comments_link": "",
   "always_display_attachments": false,
   "author": "LastWeekTonight",
   "score": 0,
   "note": null,
   "lang": ""
  }
```

This info can be used in `ttrss_api.cpp::fetch_feed`, e.g.:
```c++
bool marked = item_obj["marked"];
if (marked) {
	item.labels.push_back("ttrss:starred");
}
```

In this case, we set an newsboat label... This would be something to discuss: How do you want to use the "starrred" information within newsboat? Is a tag sufficient?

But this level of integration is a bit lagging, due to my long-term favorite #16 . By default, we only fetch a limited number of feed items from the remote, so we will not notice changes to those flag-values for older articles, so this information will only be partial and the "starred"-states will diverge over time. I don't know of an efficient way to fetch those informations in bulk...

**Comment by araujo0205** [comment]
I using tt-rss in two devices, but the starred is not sync, the newsboat only send with `ttrss-flag-star`.
if I star in one device the other don`t see the new starred article.

**Comment by Allaman** [comment]
Is there any news regarding this issue?

**Comment by Minoru** [comment]
Nope. If you want to move this forward, you can check if TT-RSS API provides a way to find out which items got starred lately. It was apparently impossible two years ago (https://github.com/newsboat/newsboat/issues/270#issuecomment-420098961), but maybe something changed? If not, consider filing an issue with TT-RSS to add such an API, because without it, Newsboat can't really implement this synchronization (see the linked comment for explanation).

**Comment by Allaman** [comment]
There is an endpoint providing this data respectively a filter to the know getHeadlines. See following request which returns my starred articles.

```
GET /api/ HTTP/1.1
Host: ttrss.example.com
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:85.0) Gecko/20100101 Firefox/85.0
Accept: text/json;charset=UTF-8
Accept-Language: en-US,de-DE;q=0.5
Accept-Encoding: gzip, deflate
Content-Type: text/json;charset=UTF-8

{
  "sid":"foobar",
  "op":"getHeadlines",
  "feed_id":"-1",
  "view_mode":"unread"
}
```

- sid is the session id returned after successfully logged in
- feed_id -1 only matches starred articles
- view_mode unread only returns unread items

[Here](https://git.tt-rss.org/git/tt-rss/wiki/ApiReference) is the according documentation.

Unfortunately, I am by no means a C++/Rust developer so I cannot contribute to the code :(

**Comment by Minoru** [comment]
Well, yes, we're using that exact API endpoint, but as explained in https://github.com/newsboat/newsboat/issues/270#issuecomment-420098961, we're only fetching a certain number of feed items. Fetching the whole feed could take a long time. Thus if a user stars an old article, Newsboat won't know about it because that article is not part of the (limited) response.

One way around that is for TT-RSS to provide a way to order items by *when* they were starred — what's what I meant by "an endpoint that lists what items have been starred lately".

**Comment by Allaman** [comment]
That's a pity because for my use case the API would be sufficient as I use starred articles for my filter to mark possible important articles and after checking them I mark them as unread and forget about them :D For storing actual items I use pocket but I guess workflows may differ...

**Comment by noctux** [comment]
Maybe some additional information:

The official API-Reference is [here](https://git.tt-rss.org/fox/tt-rss/wiki/ApiReference).

Within the client, we currently fetch a list of using the pair of `getCategories` plus `getFeeds`, and then retrieve the contents of each feed using `getHeadlines`. As indicated, this APIcall is capped at 200 entries, but the API *could* be paged, which we do not do (to me, subjectively, reloadAll is already slow enough, and the question would be how far to page).

As far as I can see, there is no support to get items that were updated after some specific timestamp or logical time, the only thing I've found was `since_id`, which is only useful to retrieve new items, but does not include old ones. If I've missed something, I would be happy for any pointers.

The official client does not have that issue: It is an online first design which, by default, does not cache any local state, and enabling/disabling offline mode is a very timeconsuming operation that diffs all state against the server.

> One way around that is for TT-RSS to provide a way to order items by when they were starred — what's what I meant by "an endpoint that lists what items have been starred lately".

There is, at least, an option to retrieve a list of all starred items, that can be queried. In theory, we could retrieve all starred items with each sync operation, and then diff locally stared and remotely starred items, and then sync only that feature. However, depending on the number of items starred, this, once more, sounds pretty expensive, and I'm not sure if thats the way to go, especially as this is only a very local fix: read state, for instance, would still remain out of sync... I'm not sure whether we should go down that particular road...

**Comment by Minoru** [comment]
It just occurred to me to look at other remote APIs for inspiration :) Turns out most of them have a `REMOTE-min-items` setting that dictates how many *pages* Newsboat will fetch on each update. So let's just do that with TT-RSS, too?

**Comment by Allaman** [comment]
As I workaround I had the idea to create a filter in TTRSS to add a special tag to all my articles that get starred allowing me to query this particular tag in newsboat to create my starred articles view. 
Unfortunately, it looks like there is no mapping between TTRSS tags and newsboat tags. Only TTRSS categories are mapped to newsboat tags. Am I right?

**Comment by Minoru** [comment]
@Allaman That's right, folders/categories are mapped into Newsboat tags. Alas, Newsboat doesn't have enough categorization options to fully support everything that TT-RSS supports.
